x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/writer/AutodetectControlMsgWriter.java

```diff
@@ -204,37 +224,44 @@ public void writeCategorizationStopOnWarnMessage(boolean isStopOnWarn) throws IO
 
     public void writeUpdateDetectorRulesMessage(int detectorIndex, List<DetectionRule> rules) throws IOException {
         StringBuilder stringBuilder = new StringBuilder();
-        stringBuilder.append(UPDATE_MESSAGE_CODE).append("[detectorRules]\n");
-        stringBuilder.append("detectorIndex=").append(Integer.toString(detectorIndex)).append("\n");
-
-        stringBuilder.append("rulesJson=");
+        stringBuilder.append(UPDATE_MESSAGE_CODE);
 
         try (XContentBuilder builder = JsonXContent.contentBuilder()) {
-            builder.startArray();
-            for (DetectionRule rule : rules) {
-                rule.toXContent(builder, ToXContent.EMPTY_PARAMS);
-            }
-            builder.endArray();
+            builder.startObject();
+            builder.field(DETECTOR_RULES).startObject();
+            builder.field(DETECTOR_INDEX, detectorIndex);
+            builder.field(CUSTOM_RULES, rules);
+            builder.endObject();
+            builder.endObject();
             stringBuilder.append(Strings.toString(builder));
         }
-
         writeMessage(stringBuilder.toString());
     }
 
     public void writeUpdateFiltersMessage(List<MlFilter> filters) throws IOException {
+
         StringBuilder stringBuilder = new StringBuilder();
-        stringBuilder.append(UPDATE_MESSAGE_CODE).append("[filters]\n");
-        new MlFilterWriter(filters, stringBuilder).write();
+        stringBuilder.append(UPDATE_MESSAGE_CODE);
+        try (XContentBuilder jsonBuilder = JsonXContent.contentBuilder()) {
+            jsonBuilder.startObject().field(MlFilter.RESULTS_FIELD.getPreferredName(), filters).endObject();
+            String msg = Strings.toString(jsonBuilder);
+            stringBuilder.append(msg);
+        }
         writeMessage(stringBuilder.toString());
     }
 
     public void writeUpdateScheduledEventsMessage(List<ScheduledEvent> events, TimeValue bucketSpan) throws IOException {
         StringBuilder stringBuilder = new StringBuilder();
-        stringBuilder.append(UPDATE_MESSAGE_CODE).append("[scheduledEvents]\n");
-        if (events.isEmpty()) {
-            stringBuilder.append("clear = true\n");
```

Comment:
So if we now have an empty list of events, c++ will receive an empty `events` array. Will this clear any existing events?

Reply:
Yes, the C++ code has been adjusted such that an empty events array, generating the message `u{"events":[]}` will clear out any existing events.